### PR TITLE
TFP: haku near-winner replication seed=123

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Replicate the haku near-winner TFP config (best_val=0.002466, best_test=0.002393) with seed=123 to sample the result distribution across seeds. Part of a 4-seed sweep (seeds 0, 7, 42, 123) to determine whether the config reliably lands below the historical target of field_mse < 0.002383.

## Dataset

`tandemfoil_paper` — paper-faithful Experiment 4 tasks, normalized full-field MSE.

## Training Command

```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion \
  --lr 1e-4 \
  --cosine-t-max 20 \
  --grad-clip 0.5 \
  --weight-decay 1e-2 \
  --model-slices 64 \
  --model-layers 3 \
  --model-hidden-dim 192 \
  --model-heads 3 \
  --enable-fourier \
  --ema-decay 0.999 \
  --epochs 999 \
  --seed 123 \
  --wandb_group tfp-haku-replication
```

## Baseline

- **best_val field_mse:** 0.002466 (haku W&B y790y0dv)
- **best_test field_mse:** 0.002393
- **Historical target:** 0.002383
- **Kill gates:** field_mse > 0.006 at ep220, > 0.004 at ep350, no improvement for 160 epochs

## Expected Outcome

Confirm stability of the haku config at seed=123. A result at or below 0.002383 would beat the historical target. Report best_val and best_test (from best-val checkpoint) field_mse alongside the W&B run ID.

## Instructions

1. Run the training command above exactly as written.
2. Monitor: kill if field_mse > 0.006 at ep220, > 0.004 at ep350, or no improvement for 160 consecutive epochs.
3. Report best_val/field_mse, best_test/field_mse (evaluated from best-val checkpoint), and the W&B run ID in a PR comment.
4. Include the full W&B run URL so the advisor can verify metrics independently.

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: **CONFIG ERROR — CRITICAL.** Your current command is MISSING all physics features (--enable-te-coord-frame, --enable-cp-panel, --enable-cp-panel-tandem-only, --asinh-pressure, --residual-prediction, --enable-pressure-prior-addition). Without these, field_mse cannot reach below ~0.014 — the kill gate will terminate the run.

**When your current run finishes (or is killed), re-run with this corrected full TFP champion config:**

```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion \
  --lr 1e-4 \
  --cosine-t-max 10 \
  --grad-clip 0.5 \
  --weight-decay 1e-2 \
  --model-layers 3 \
  --model-hidden-dim 192 \
  --model-heads 3 \
  --enable-fourier \
  --enable-te-coord-frame \
  --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure \
  --residual-prediction \
  --enable-pressure-prior-addition \
  --ema-decay 0.999 \
  --epochs 999 \
  --seed SEED \
  --wandb-group tfp-champion-replication \
  --kill-if-best-val-above 220:0.006 \
  --kill-if-best-val-above 350:0.004 \
  --kill-if-no-improvement 160:0.001
```

Replace SEED with your assigned seed. Baseline to beat: val_primary/field_mse = 0.002344 (PR #3056).

**Comment by tcapelle** [comment]
STUDENT yuji:

## Results

**Run killed at epoch 393 — ep350 kill gate triggered.**

### Key Metrics

| Metric | Value | Baseline (haku y790y0dv) | Delta |
|--------|-------|--------------------------|-------|
| best_val field_mse | 0.004257 | 0.002466 | +72.6% (worse) |
| best_test field_mse | N/A (killed) | 0.002393 | — |
| best_epoch | 387 | — | — |
| historical target | — | 0.002383 | — |

### Kill Gate Log

- **ep220:** best=0.005621 < 0.006 ✅ passed
- **ep350:** best=0.004376 > 0.004 ❌ **triggered — run killed**

### Training Trajectory

| Epoch | field_mse | Best | Best Epoch |
|-------|-----------|------|------------|
| 1 | 0.5171 | 0.5171 | 1 |
| 50 | 0.0233 | 0.0233 | 50 |
| 100 | 0.0104 | 0.0104 | 100 |
| 200 | 0.0059 | 0.0059 | 200 |
| 220 | 0.0056 | 0.0056 | 215 |
| 300 | 0.0048 | 0.0046 | 284 |
| 350 | 0.0044 | 0.0044 | 349 |
| 387 | 0.0043 | 0.0043 | 387 |
| 393 | 0.0045 | 0.0043 | 387 |

### Experiment Details

- **Command:** `cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil_paper --optimizer lion --lr 1e-4 --cosine-t-max 20 --grad-clip 0.5 --weight-decay 1e-2 --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --ema-decay 0.999 --epochs 999 --seed 123 --wandb-group tfp-haku-replication --agent yuji --wandb-name "yuji/tfp-haku-seed123"`
- **W&B run ID:** f9vsnv6f
- **W&B run URL:** https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/f9vsnv6f
- **Peak VRAM:** ~56.8 GB (RTX PRO 6000 Blackwell)
- **GPU:** CUDA:0

### What happened

Seed 123 significantly underperformed the haku near-winner baseline (0.004257 vs 0.002466, +72.6% worse). The run showed consistent descent through ~393 epochs but was converging to a much shallower basin. The ep350 kill gate (best > 0.004) was triggered with best=0.004376, and the run was terminated.

The convergence profile was smooth — no divergence, no NaN, no instability. The model simply settled into a weaker optimum at this seed. The gap between this seed and the haku run is large (>72%), suggesting the original haku result at seed=42 (default) may have been a lucky seed for this particular configuration.

### Suggested follow-ups

- The 4-seed sweep (seeds 0, 7, 42, 123) should reveal whether the haku config's sensitivity to seed is systematic. If 3+ seeds fail the kill gate, the config may not be reliably sub-0.0024.
- Consider whether T_max=20 + lr=1e-4 has narrower basins of attraction than the champion config (T_max=10 + lr=1.25e-4), making it more seed-dependent.

**Comment by tcapelle** [comment]
ADVISOR: This run used the broken config (missing physics features) — the 0.004257 result is expected without TE/Cp/asinh/residual features. Please re-run with the CORRECTED full TFP champion config I posted earlier:

```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion \
  --lr 1e-4 \
  --cosine-t-max 10 \
  --grad-clip 0.5 \
  --weight-decay 1e-2 \
  --model-layers 3 \
  --model-hidden-dim 192 \
  --model-heads 3 \
  --enable-fourier \
  --enable-te-coord-frame \
  --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure \
  --residual-prediction \
  --enable-pressure-prior-addition \
  --ema-decay 0.999 \
  --epochs 999 \
  --seed 123 \
  --wandb-group tfp-champion-replication \
  --kill-if-best-val-above 220:0.006 \
  --kill-if-best-val-above 350:0.004 \
  --kill-if-no-improvement 160:0.001
```

Baseline to beat: val_primary/field_mse = 0.002344 (PR #3056). Report both val and test field_mse from best checkpoint.

**Comment by tcapelle** [comment]
STUDENT yuji:

## Results — Corrected TFP Champion Config (seed=123)

**Run killed at epoch 145 — manual termination due to catastrophic training instability and imminent no-improvement gate (ep163).**

### Key Metrics

| Metric | Value | Baseline (PR #3056) | Historical target |
|--------|-------|---------------------|-------------------|
| best_val field_mse | 6307.66 | 0.002344 | 0.002383 |
| best_test field_mse | N/A (never improved past ep3) | — | — |
| best_epoch | 3 | — | — |

### Kill Gate Status

- **ep220 gate (>0.006):** Would have triggered — best_val = 6307.66 >> 0.006
- **ep350 gate (>0.004):** Would have triggered
- **No improvement gate (160ep):** Would have triggered at ep163 — best never improved past epoch 3

### Training Trajectory

| Epoch | val/field_mse | loss | Notes |
|-------|--------------|------|-------|
| 1 | 7.82e19 | 0.857 | Start |
| 3 | 6,308 (BEST) | 0.369 | Best val — never beaten |
| 10 | 8.6e7 | 0.178 | Early oscillations |
| 25 | 6.7e10 | 0.098 | Peak instability, EMA at target |
| 50 | 1.6e8 | 0.067 | Oscillating wildly |
| 75 | 407K | 0.061 | Val floor descending |
| 92 | 91K | 0.054 | New low |
| 100 | 346K | 0.050 | Oscillating 90K-660K |
| 114 | 27K (new low) | 0.048 | Best floor since ep3 |
| 130 | 398K | 0.049 | Still oscillating |
| 138 | 587K | 0.051 | Pre-instability |
| 139 | 91.6M | **0.226** | **Catastrophic instability — loss 4.4x spike** |
| 142 | 3.7M | **0.317** | Loss near epoch-1 levels |
| 145 | 88.8M | 0.131 | No recovery — killed |

### Experiment Details

- **Command:** `cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil_paper --optimizer lion --lr 1e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --ema-decay 0.999 --epochs 999 --seed 123 --wandb-group tfp-champion-replication --agent yuji --wandb-name "yuji/tfp-champion-seed123"`
- **W&B run ID:** sr55u4qu
- **W&B run URL:** https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/sr55u4qu
- **Peak VRAM:** ~14.4 GB (RTX PRO 6000 Blackwell)
- **GPU:** CUDA:0

### What happened

The corrected TFP champion config with full physics features (--enable-te-coord-frame, --enable-cp-panel, --enable-cp-panel-tandem-only, --asinh-pressure, --residual-prediction, --enable-pressure-prior-addition) produced catastrophically unstable training on tandemfoil_paper at seed=123.

**Key observations:**

1. **Scale mismatch:** The best val field_mse (6307 at ep3) is ~2.7 million times above the baseline target (0.002344). For comparison, the broken config (without physics features) reached field_mse=0.0104 at ep100. The corrected config's field_mse at ep100 was 346,000 — 33 million times worse.

2. **Persistent val oscillations:** Despite training loss falling steadily (0.857→0.048 over 138 epochs), the val field_mse oscillated wildly between 27K and 2M, never beating the epoch 3 best of 6308. The volume_mse component drove most of the instability.

3. **Catastrophic instability at ep139:** At epoch 139, training loss spiked from 0.051 to 0.226 (4.4x), then progressively worsened to 0.317 at ep142 — nearly back to epoch 1 levels. Grad norms spiked from ~9 to 27, with all steps clipped. The model never recovered.

4. **Possible root cause:** The combination of residual_prediction + asinh_pressure + pressure_prior_addition may be incompatible with tandemfoil_paper, or these features may amplify numerical instability in the volume prediction for this dataset. Note that `cp_panel_prior_index` returns None for tandemfoil_paper (it's not in the supported list), so pressure_prior_addition is silently disabled, yet the other physics features still alter the model's input/output representation.

### Suggested follow-ups

- **Investigate compatibility:** The TFP champion config was likely tuned on `tandemfoil` (not `tandemfoil_paper`). The physics features (especially residual_prediction, asinh_pressure) may need different hyperparameters or may not be appropriate for the paper-faithful dataset.
- **Run without physics features:** The original haku config (without physics features) produced field_mse=0.0104 at ep100 — much more reasonable. The physics features may hurt rather than help on tandemfoil_paper.
- **Check other seeds:** If seeds 0, 7, 42 show similar scale mismatch with the corrected config, the issue is config-dataset incompatibility, not seed sensitivity.
- **Cosine T_max interaction:** The catastrophic instability at ep139 coincided with a cosine annealing boundary (T_max=10, ep139 mod 10 = 9). The warm restart may be too aggressive for this unstable training regime.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3292 because ADVISOR: Closing — TFP seed=123 with full-physics config catastrophically unstable (val field_mse oscillated 27K-2M, never converged). Killed at ep145. The residual_prediction + asinh_pressure + physics features combination is numerically unstable on TFP at this seed. Yuji reassigned..
